[dynamo] Refactor install_global_once, remove usages of install_global_unsafe

zou3519 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #118100

We split install_global_once into two APIs:
- `install_global_by_id(prefix, value) -> name`: installs a global if it hasn't
been installed yet
- `install_global(prefix, value) -> name`: always installs the global (and
  generates a unique name for it)

Then, we refactor most callsites of `install_global_unsafe` to one of
the previous. Some callsites cannot be refactored because we create the
global name first, do a lot of stuff with it, and then install it.

This fixes more test flakiness.

Test Plan:
- Existing tests; I can't reliably repro the flakiness

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng

zou3519 committed:
[dynamo] Remove install_global_unsafe from tx.store_global_weakref

This was leading to some more flakiness. Also cleaned up the code a
little.

Test Plan:
- Existing tests; I can't reliably repro the flakiness

[ghstack-poisoned]

zou3519 committed:
Update on "[dynamo] Remove unsafe usages of install_global_unsafe"


This was leading to some more flakiness. Also cleaned up the code a
little.

Test Plan:
- Existing tests; I can't reliably repro the flakiness

[ghstack-poisoned]

ezyang commented on torch/_dynamo/symbolic_convert.py:
```diff
@@ -2187,6 +2190,7 @@ def create_call_resume_at(self, inst):
         if new_code.co_freevars:
             cg.make_function_with_closure(name, new_code, True, stack_len)
         else:
+            # This is safe: we pre-generate a unique name
```
Err, why? It looks to me like `__resume_at_20` could easily clobber other resumes?

zou3519 replied:
`unique_id` returns a unique number every time it is called. This number is never reset. `__resume_at_20` will only be generated once per python program.

ezyang commented on torch/_dynamo/variables/torch_function.py:
```diff
@@ -156,28 +156,30 @@ def from_tensor_var(cls, tx, tensor_var, class_type, torch_function_fn):
-        return f"__subclass_{self.class_type.__name__}_{id(self.class_type)}"
+    def global_mangled_class_name(self, tx):
+        compile_id = tx.output.compile_id
+        return f"__subclass_{self.class_type.__name__}_{id(self.class_type)}_c{id}"
```
@mlazos could you review this part

ezyang commented on torch/_dynamo/variables/optimizer.py:
```diff
@@ -62,7 +62,8 @@ def call_method(
                 # if the optimizer object dies
-                tx.store_global_weakref(self.get_global_name(), self.value)
+                mangled_name = f"__optimizer_{id(self.value)}"
+                tx.store_global_weakref(mangled_name, self.value)
```
I don't quite understand what the rule for the refactor was here

zou3519 replied:
This is the only use of self.get_global_name(), so I inlined it here.

zou3519 replied:
Not applicable anymore

ezyang reviewed (approved):
Seems OK?

zou3519 commented on torch/_dynamo/output_graph.py:
```diff
@@ -884,7 +884,7 @@ def append_prefix_insts():
             rand_fn_name = unique_id("__gen_rand_values")
             rand_fn = disable(_get_gen_rand_values_fn(tx.random_calls))
-            self.install_global_unsafe(rand_fn_name, rand_fn)
+            rand_fn_name = self.install_global_once(rand_fn_name, rand_fn)
```
The problem with (2) is that the common pattern is to generate the name, then install the global, and then use the name of the global to do something else (e.g. creating a source with it). As a result, we need some way to return the generated name from the `install_global` function.

zou3519 replied:
I split install_global_once into two APIs. One of them does the unique_id automatically. Both of these APIs take in a "prefix" (instead of a "name"), so hopefully it'll be less footgun-y.

zou3519 committed:
Update on "[dynamo] Refactor install_global_once, remove usages of install_global_unsafe"


We split install_global_once into two APIs:
- `install_global_by_id(prefix, value) -> name`: installs a global if it hasn't
been installed yet
- `install_global(prefix, value) -> name`: always installs the global (and
  generates a unique name for it)

Then, we refactor most callsites of `install_global_unsafe` to one of
the previous. Some callsites cannot be refactored because we create the
global name first, do a lot of stuff with it, and then install it.

This fixes more test flakiness.

Test Plan:
- Existing tests; I can't reliably repro the flakiness

[ghstack-poisoned]

zou3519 commented:
@pytorchbot merge -f "passed Dynamo CI"
